ddtrace/tracer/payload_test.go

```diff
@@ -43,7 +44,6 @@ func TestPayloadIntegrity(t *testing.T) {
 			want.Reset()
 			err := msgp.Encode(want, lists)
 			assert.NoError(err)
-			assert.Equal(want.Len(), p.size())
```

Comment:
Well actually this test is not really relevant since we already compare the actual msgpack encoded outputs. Moreover the current size method provides an upper bound on the encoded payload size, not the exact encoded payload size.

Reply:
That's true, but since this method (`(*payload).size`) is such a key element to triggering a flush to the transport, there should be a test asserting that its return value is somewhat close to what it should be.

Reply:
Hi @gbbr,
Digging a bit into the trace.MsgSize() method provided by msgp, the actual encoded size of the payload for the test cases in payload_test.go is usually around 70% of the size returned by the new payload.size() method. Since it is an upper bound of the size the encoded payload could potentially have, I would not find it to be an issue (no risk of over-allocating memory), but it makes it less obvious to write a test case for this method. What do you think the passing condition for this test should be? Do you think this discrepancy could be an issue ?

Reply:
To give more details: this 30 % discrepancy is only true for traces with only very small spans with no metadata. I tested with a payload filled with traces populated with spans with 1KB of metadata each, the size estimate because then much more accurated, with à 5% error. So to conclude the error starts out at 30%, and drops towards 0% as spans grow in size.

Reply:
I think the only potential solution here is somewhere along the lines of:

* Knowing that the size is around 66% (based on my investigations) change [`payloadSizeLimit`](https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/tracer.go#L57) to be `payloadMaxLimit / 4`.
* Write a test generating a large variety of payloads (no larger than 10MB) and assert for all of them that the resulting `size()` is between 60-100% of the actual size.
* Update the comment of `payload.size()` and the `msgSize` field to specify that it's an estimation and not the actual size. Be explicit saying it's an "upper bound estimate" and that it's generally X% of the size.

I'd start with the test to get a better idea of the %.

This has to be tested well though and seems like a big change. I'm not sure if this is a change that is worthwhile at this point and I'm wondering if it's not better to just simply drop the "retry policy" and reset the payload every time the transport fails (i.e. agent restarts, etc.). This will partially solve your problem with the invalid stale payload.

Reply:
However I really like this solution and I like postponing the encoding part to the transport, as opposed to doing it on the fly as spans finish.

Reply:
@gbbr it is the reverse, the actual encoded size is currently between 66 and 100% of the value returned by payload.size() (which is an upper bound of the actual encoded size).
Therefore I think setting `payloadSizeLimit` to `int(float64(payloadMaxLimit)*0.8)` would be appropriate. What do you think about this?

Regarding the two last points, I will implement them.